### ISSUE TITLE
Export internal types such as ExtractState

### ISSUE THREAD

**Issue by underfisk** [pmndrs/zustand#1456]
## Summary
Currently, when consuming the new v4 context-based API it can become verbose and in order to facilitate the creation of a context-based store in my current project I do need to re-use a few types and provide more type safety at the application level. 
Not exposing the internal types could be international but for library authors or anyone that's actually consuming and enhancing local code, it can become challenging and have to perform a manual copy.

My request would simply be exposing those types because at some point they are handy to quickly override such as:
```typescript
const myStore = create(middleware1(...))
````
The following Channing can produce something like
````typescript
UseBoundStore<Write<StoreApi<State>, StoreSubscribeWithSelector<State>>>
````

It would be good if we could gather those types or have some sort of alias and simply use those and they are helpful when it comes to creating local store refs

## Link to reproduction
Attempting to import types like `ExtractType`, `StoreSubscribeWithSelector`, or `Write` will result in an error

## Check List

Please do not ask questions in issues.

- [X] I understand this is not a question.

Please fill this template if you're filling an issue regarding TypeScript.

- [ ] I've added a link to a [typescript playground](https://www.typescriptlang.org/play) or [codesandbox](https://codesandbox.io) with a minimal reproduction. Or some other way of reproduction if these two don't suffice.
- [X] I've read the [typescript guide](https://github.com/pmndrs/zustand/blob/main/docs/guides/typescript.md), in particular that `create` is to be used as `create<T>()(...)` and not `create<T>(...)`.

**Comment by dai-shi** [comment]
Exporting `ExtractState` may make sense. We can do that.
What's your use case?

---

`Write` and `WithSomething` are internal types and people should never use it. If we export them, people would just misuse them.

It feels like the zustand typing is just very hard to understand for many developers, and we fail to educate the usage very well.

So, If I understand your case correctly, it should be something like this.
https://tsplay.dev/wRz5Yw
```ts
import create, { UseBoundStore, Mutate, StoreApi } from "zustand";
import { subscribeWithSelector } from "zustand/middleware";

type MyState = {
  count: number;
  inc: () => void;
};

type UseMyStore = UseBoundStore<
  Mutate<StoreApi<MyState>, [["zustand/subscribeWithSelector", never]]>
>;

const useMyStore: UseMyStore = create<MyState>()(
  subscribeWithSelector((set) => ({
    count: 0,
    inc: () => set((state) => ({ count: state.count + 1 })),
  }))
);
```
Does it look something you were trying?

See also #930 discussion.

**Comment by underfisk** [comment]
@dai-shi Thanks for getting back to me so fast. For `Write` and other internal types, you're right i probably don't need them and the following example helps, it allows us to build a re-usable type with `subscribeWithSelector` which fixes one problem. 
The other problem might occur with the function overloading types that we don't have access to which i ended up having a manual copy.

## What might be the use for the overloading types?
Let's consider the following:
```typescript
const MyComponent = () => {
    const api = useContext(ZustandStoreContext)
    const data = useStore(api)
    const selectedData = useStore(api,state => state.whatever)
}
```
It ends up being how we consume the context but if you want to build a re-usable pattern you'll need to abstract a few steps such as providing the API if you already have it
```typescript
type UseStoreFunc<T, ReturnedData> = {
  (): ExtractState<T>;
  (selector: (state: ExtractState<StoreApi<T>>) => ReturnedData, equalityFn?: (a: ReturnedData, b: ReturnedData) => boolean): ReturnedData
}

type Params<S, R> = Parameters<UseStoreFunc<S, R>>
export const createZustandContext = <State,>() => {
  const { Provider } = createContext<Store<State>>({} as Store<State>)
  let storeApi: Store<State>

  const useStoreProxy = <ReturnValue = State>(...params: Params<State, ReturnValue>) => {
    return useStore(storeApi, params?.[0] ?? ((state: State) => state as State) as any, params?.[1])
  }

  const useStoreApi = () => {
    return storeApi
  }

  const initializeStore = (initializer: StateCreator<State>) => {
    storeApi = create(initializer)
    return storeApi
  };

  return {
    initializeStore,
    Provider,
    useStore: useStoreProxy,
    useStoreApi
  };
};
```

The challenge here was to be able to already provide the API so that the consumer would continue using the API the same way as a global store (trying to avoid the boilerplate)
Let me know if there's anything else i can provide but essentially the use case here is mostly introducing a generic helper function that could alleviate the boilerplate when dealing with context-based stores

**Comment by dai-shi** [comment]
Yeah, function overloading is a difficult one. I don't think exporting types helps it anyway. I also think it's a rare case anyways. (And, there should be easier workarounds for TS beginners to avoid function overloading.)

Here's what I've got: https://tsplay.dev/mbK0dW

**Comment by underfisk** [comment]
@dai-shi I wanted to have the most type of safety so that once we upgrade (if there is any breaking change) to easily catch those but I think your solution can probably alleviate and temporarily solve the problem that I'm facing.
Thanks for the responses! Feel free to close if there's no action item out of this request
